### PR TITLE
feat(aem): add support for generic model backed by a model.json

### LINKED ISSUE #12: Automatic model modules loading
the current set of a _imported_ modules for the models are hardcoded,
it should be possible to include import them automatically (during compile time)

see GenericModel

### PR DESCRIPTION
**Pull request by tripodsan**
## What I did
- moved the `GenericModel` from POC to Aem client
- removed explicit model imports from `GenericModel`.
- improved the `GenericModel` to support use-classes, model objects and resource paths
- added new text story that uses a content-backed generic model.

## Notes:
The models used by the HTL scripts need now to be declared explicitly in the stories. on the one hand this is very powerful, as the story can now influence which model classes or model content to use. but for the most cases, this is just extra work. 

In the future, the model use-classes (similar to the component `.content.xml`) should be registered automatically during build time.

fixes #12

### CODE REVIEW

**Comment by CodeByAlex** [comment]
Great idea to supply different options to the user 👍

File: app/aem/src/client/preview/helpers/GenericModel.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/**
+ * Generic model implementation that either instantiates a use-class or mocks a Sling Model class.
+ */
+class GenericModel {
+  constructor(id) {
+    this.id = id;
+  }
+
+  async use(context) {
+    console.log(context);
```
Comment (CodeByAlex) [inline app/aem/src/client/preview/helpers/GenericModel.js]: Can probably remove this

File: examples/aem-kitchen-sink/core/wcm/components/accordion/accordion.stories.js [hunk 2]
```diff
@@ -6,6 +6,13 @@ export default {
 
 export const Accordion = () => {
   return {
+    // models used to render this component. the model can either be a proper use-class, a
```
Comment (CodeByAlex) [inline examples/aem-kitchen-sink/core/wcm/components/accordion/accordion.stories.js]: Can we add this to the documentation in the readme?

File: examples/aem-kitchen-sink/core/wcm/components/text/text.stories.js [hunk 3]
```diff
@@ -42,3 +51,21 @@ export const RichText = () => {
     noDecoration: true
   };
 };
+
+export const TextWithModelJson = () => {
+  return {
+    // example of using a generic model, backed with content
+    models: {
+      'com.adobe.cq.wcm.core.components.models.Text': '/'
```
Comment (CodeByAlex) [inline examples/aem-kitchen-sink/core/wcm/components/text/text.stories.js]: Is the comment referring to the “/“?

File: examples/aem-kitchen-sink/core/wcm/components/accordion/accordion.stories.js (hunk 2 above)
Comment (tripodsan) [reply examples/aem-kitchen-sink/core/wcm/components/accordion/accordion.stories.js]: ok. I added it to the readme.

File: examples/aem-kitchen-sink/core/wcm/components/text/text.stories.js (hunk 3 above)
Comment (tripodsan) [reply examples/aem-kitchen-sink/core/wcm/components/text/text.stories.js]: yes

File: app/aem/src/client/preview/helpers/GenericModel.js (hunk 1 above)
Comment (tripodsan) [reply app/aem/src/client/preview/helpers/GenericModel.js]: ok. removed it

**Comment by CodeByAlex** [comment]
Do you think you could show a story with inlined model json (not in a file) as an example?

File: examples/aem-kitchen-sink/core/wcm/components/text/text.stories.js (hunk 3 above)
Comment (tripodsan) [reply examples/aem-kitchen-sink/core/wcm/components/text/text.stories.js]: I added a better example, using a 'real' path.

**Comment by tripodsan** [comment]
> Do you think you could show a story with inlined model json (not in a file) as an example?

ok. added one.

**Comment by CodeByAlex** [comment]
@tripodsan, with your solution, is it still requiring that teams replicate their models as js?

**Comment by tripodsan** [comment]
> @tripodsan, with your solution, is it still requiring that teams replicate their models as js?

only if they have special logic in them... but if the models are pure _getters_ to the underlying resource content (i.e. if the exported `model.json` provides a representation of the native types for the scripts, then you don't need a use-class.

the following wouldn't work with a pure mode.json dump:
- computed properties: eg if the use-class exports a `fullname()` computed property, knobbing the underlying content (`firstname`, `lastname`) wouldn't change the `fullname()` property in the `model.json` obviously.
- complex types: some use-classes return complex types, that themselves have logic on them. e.g. a list component could return an iterator of referenced page-objects, which in turn compute the title-text smartly....
- non-model properties: it is possible to have a use-class that exposes a property, which isn't exported to the model.json, but still accessible in the script.

**Comment by CodeByAlex** [comment]
Thanks for the explanation 👍 

Okay last thing :)

If we can include the model.json as a file, can we assume that we could write out the json as a part of the story config instead of as a part of a separate file for really simple use cases?

Btw the examples of when you might use the different options seem to help and may help the user best navigate which to use for their scenario - not a blocker but at some point, could you add something similar to the readme?

LGTM!

File: README.md [hunk 4]
```diff
@@ -24,16 +24,10 @@ To build and test out this project complete the following:
...
 - better error reporting (add file, line, col): eg: `Error: Error: mismatched input 'in' expecting {'}', '@'}`
-- add support for function getters and java-like getters. e.g. `getText` for ${xyz.text}
-- repeat variable not available in attributes of same element. eg:
-  ```
-   <div data-sly-repeat.item="${accordion.items}" data-cmp-expanded="${accordion.expandedItems[item.name] ? true : false}">
-  ```
 
 ## Usage
-See [example](https://github.com/storybookjs/aem/blob/master/examples/aem-kitchen-sink/components/text/text.stories.js):
+See [example](./examples/aem-kitchen-sink/core/wcm/components/text/text.stories.js):
```
Comment (CodeByAlex) [inline README.md]: Nice 👍 noticed this error today - glad you caught it

**Comment by tripodsan** [comment]
> If we can include the model.json as a file, can we assume that we could write out the json as a part of the story config instead of as a part of a separate file for really simple use cases?

yes. see:

https://github.com/storybookjs/aem/pull/26/files#diff-256959bac2964ee9519c1e37f8111c42R62-R65

I'll try to make the distinction between the different ways to use the models more clear...

**Comment by CodeByAlex** [comment]
@tripodsan i must have missed that example - awesome!!

File: README.md [hunk 5]
```diff
@@ -99,6 +105,83 @@ require('./clientlibs/site/js/accordion.js');
...
+      ':items': {
+        text0001: {
+          text: 'Hello, world',
+          isRichText: false,
+        }
+      }
+    }
+```
+
+**Note**: In the future, the content path would be derived from the current resource path, so that it will be very simple to render a complex component. It might even become the default (e.g. merge with (1) above.
```
Comment (tripodsan) [inline README.md]: @CodeByAlex  After I wrote this, I realized that we can collapse (1) and (3) here.... I will work on in after this PR is merged.

Comment (CodeByAlex) [reply README.md]: Great documentation here btw!

**Pull request merged**